go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -165,59 +178,138 @@ func (oc *Controller) updateEgressFirewallWithRetry(egressfirewall *egressfirewa
+		for _, joinSwitch := range joinSwitches {
+			err := createACLRule(defaultStartPriority-rule.id, match, action, ef.namespace, joinSwitch)
+			if err != nil {
+				return err
+			}
 		}
-		uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
-			"--columns=_uuid", "find", "ACL", match, "action="+action,
-			fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace))
+	}
+	return nil
+}
+
+func createACLRule(priority int, match, action, namespace, joinSwitch string) error {
+	uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=_uuid", "find", "ACL", match, "action="+action,
+		fmt.Sprintf("external-ids:egressFirewall=%s", namespace))
 
+	if err != nil {
+		return fmt.Errorf("error executing find ACL command, stderr: %q, %+v", stderr, err)
+	}
+	if uuid == "" {
+		_, stderr, err := util.RunOVNNbctl("--id=@acl", "create", "acl",
+			fmt.Sprintf("priority=%d", priority),
+			fmt.Sprintf("direction=%s", fromLport),
+			match, "action="+action,
+			fmt.Sprintf("external-ids:egressFirewall=%s", namespace),
+			"--", "add", "logical_switch", joinSwitch,
+			"acls", "@acl")
 		if err != nil {
-			return fmt.Errorf("error executing find ACL command, stderr: %q, %+v", stderr, err)
+			return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
 		}
-		for _, joinSwitch := range joinSwitches {
-			if uuid == "" {
-				_, stderr, err := util.RunOVNNbctl("--id=@acl", "create", "acl",
-					fmt.Sprintf("priority=%d", defaultStartPriority-rule.id),
-					fmt.Sprintf("direction=%s", fromLport), match, "action="+action,
-					fmt.Sprintf("external-ids:egressFirewall=%s", ef.namespace),
-					"--", "add", "logical_switch", joinSwitch,
-					"acls", "@acl")
-				if err != nil {
-					return fmt.Errorf("error executing create ACL command, stderr: %q, %+v", stderr, err)
-				}
-			} else {
-				_, stderr, err := util.RunOVNNbctl("add", "logical_switch", joinSwitch, "acls", uuid)
-				if err != nil {
-					return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
+	} else {
+		_, stderr, err := util.RunOVNNbctl("add", "logical_switch", joinSwitch, "acls", uuid)
+		if err != nil {
+			return fmt.Errorf("error adding ACL to joinsSwitch %s failed, stderr: %q, %+v", joinSwitch, stderr, err)
 
-				}
-			}
 		}
 	}
 	return nil
 }
 
+func generateMatch(ipv4Source, ipv6Source string, destinations []string, ports []egressfirewallapi.EgressFirewallPort) (string, error) {
+	var src string
+	var dst string
+	if ipv4Source == "" {
+		src = "ip4.src != 0.0.0.0/0"
+	} else {
+		src = fmt.Sprintf("ip4.src == $%s", ipv4Source)
+	}
+	if ipv6Source == "" {
+		src = fmt.Sprintf("%s || ip6.src != ::/0", src)
+	} else {
+		src = fmt.Sprintf("%s || ip6.src == $%s", src, ipv6Source)
+	}
+
+	if len(destinations) == 0 {
+		return "", fmt.Errorf("there is no valid destination for an egressfirewall rule")
+	}
+	if len(destinations) == 1 {
```

Comment:
I don't get this code. Why are there fixed offsets? Shouldn't it be generic - each destination is either a v4, v6, or address set? And you could have, say, 100 arbitrary destinations?

Reply:
removed the fixed offsets